src/Razor/src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs

```diff
@@ -69,6 +80,112 @@ public RazorDocumentSyntax ParseDocument()
             }
         }
 
+        //
+        // This will be called by the code parser whenever any markup is encountered inside a code block @{}.
+        // It can either be a single line markup like @: or a tag. In this case, we want to keep track of tag nesting
+        // and add appropriate errors for any malformed cases.
+        // In addition to parsing regular tags, we also understand special "text" tags. These tags are not rendered to the output
+        // but used to render a block of text it encloses as markup. They are a multiline alternative to the single line markup syntax @:
+        //
+        // The tag stack inside a code block is different from the stack outside the block.
+        // E.g, `<div> @{ </div> }` will be parsed as two separate elements with missing end and start tags respectively.
+        //
+        public MarkupBlockSyntax ParseBlock()
+        {
+            if (Context == null)
+            {
+                throw new InvalidOperationException(Resources.Parser_Context_Not_Set);
+            }
+
+            var oldTagTracker = _tagTracker;
+            try
+            {
+                // This is the start of a new block. We don't want the current tag stack to mix with the tags in this block.
+                // Initialize a new stack.
+                _tagTracker = new Stack<TagTracker>();
+                using (var pooledResult = Pool.Allocate<RazorSyntaxNode>())
+                using (PushSpanContextConfig(DefaultMarkupSpanContext))
+                {
+                    var builder = pooledResult.Builder;
+                    if (!NextToken())
+                    {
+                        return null;
+                    }
+
+                    AcceptWhile(IsSpacingToken(includeNewLines: true));
+                    builder.Add(OutputAsMarkupLiteral());
+
+                    if (At(SyntaxKind.OpenAngle))
+                    {
+                        ParseMarkupInCodeBlock(builder);
+                    }
+                    else if (At(SyntaxKind.Transition))
+                    {
+                        ParseMarkupTransition(builder);
+                    }
+                    else
+                    {
+                        Context.ErrorSink.OnError(
```

Comment:
Ugh... this old thing is still around?

Reply:
It's not really old anymore right? It works completely over `RazorDiagnostic`

Reply:
Does it attach the diagnostics to nodes? or does it throw them in a bucket?

Reply:
It throws them in a bucket. I can't remember specifically but there are some errors that aren't node specific or span multiple nodes. Having this is useful for those cases. But I agree with you, a lot more errors that we add to the bucket can just be node specific diagnostics. I didn't do this partly because that changes baselines. (Node diagnostics don't show up in `xxx.diag.txt` but instead will go in `xxx.stree.txt`.